### PR TITLE
enhance namespace management

### PR DESCRIPTION
**Pull request by ningmingxiao**
Signed-off-by: ningmingxiao <ning.mingxiao@zte.com.cn>

### CODE REVIEW

File: cmd/nerdctl/namespace.go [hunk 1]
```diff
@@ -114,7 +125,130 @@ func namespaceLsAction(cmd *cobra.Command, args []string) error {
...
+	}
+	defer cancel()
+	opts, err := deleteOpts(cmd)
+	if err != nil {
+		return err
+	}
+	namespaces := client.NamespaceService()
+
+	for _, target := range args {
+		if err := namespaces.Delete(ctx, target, opts...); err != nil {
```
Comment (AkihiroSuda) [inline cmd/nerdctl/namespace.go:171]: We have to delete containers, images, and volumes before deletion of the namespace

File: README.md [hunk 2]
```diff
@@ -1081,6 +1084,11 @@ Usage: `nerdctl volume rm [OPTIONS] VOLUME [VOLUME...]`
 
 ## Namespace management
 
+### :nerd_face: :blue_square: nerdctl namespace create
+Create a new namespace.
+
+Usage: `namespaces create NAMESPACE [<key>=<value]`
```
Comment (AkihiroSuda) [inline README.md:1090]: `<key>=<value>` should be `--label`

File: README.md [hunk 3]
```diff
@@ -1089,6 +1097,19 @@ Usage: `nerdctl namespace ls [OPTIONS]`
 Flags:
 - `-q, --quiet`: Only display namespace names
 
+### :nerd_face: :blue_square: nerdctl namespace remove
+Remove a new namespace.
```
Comment (AkihiroSuda) [inline README.md:1101]: Not "new"

File: cmd/nerdctl/namespace.go [hunk 4]
```diff
@@ -84,10 +94,11 @@ func namespaceLsAction(cmd *cobra.Command, args []string) error {
 
 	w := tabwriter.NewWriter(cmd.OutOrStdout(), 4, 8, 4, ' ', 0)
 	// no "NETWORKS", because networks are global objects
-	fmt.Fprintln(w, "NAME\tCONTAINERS\tIMAGES\tVOLUMES")
+	fmt.Fprintln(w, "NAME\tCONTAINERS\tIMAGES\tVOLUMES\tLABELS")
```
Comment (AkihiroSuda) [inline cmd/nerdctl/namespace.go:97]: Labels should be shown in `nerdctl namespace inspect`, not `nerdctl namespace ls`

File: cmd/nerdctl/namespace.go [hunk 5]
```diff
@@ -114,7 +125,130 @@ func namespaceLsAction(cmd *cobra.Command, args []string) error {
...
+func newNamespaceRmCommand() *cobra.Command {
+	namespaceRmCommand := &cobra.Command{
+		Use:           "remove [OPTIONS] NAMESPACE [NAMESPACE...]",
+		Aliases:       []string{"rm"},
+		Short:         "Remove one or more namespaces",
+		RunE:          namespaceRmAction,
+		SilenceUsage:  true,
+		SilenceErrors: true,
+	}
+	namespaceRmCommand.Flags().BoolP("cgroup", "c", false, "delete the namespace's cgroup")
```
Comment (AkihiroSuda) [inline cmd/nerdctl/namespace.go:150]: Where is this handled?

File: cmd/nerdctl/namespace.go [hunk 6]
```diff
@@ -114,7 +125,130 @@ func namespaceLsAction(cmd *cobra.Command, args []string) error {
...
+		SilenceErrors: true,
+	}
+	namespaceCreateCommand.Flags().BoolP("cgroup", "c", false, "delete the namespace's cgroup")
+	return namespaceCreateCommand
+}
+
+func namespaceCreateAction(cmd *cobra.Command, args []string) error {
+	if len(args) == 0 {
+		return errors.New("please specify a namespace")
+	}
```
Comment (AkihiroSuda) [inline cmd/nerdctl/namespace.go:201]: Use cobra.Command.Args to limit the arg num

File: cmd/nerdctl/namespace.go [hunk 7]
```diff
@@ -114,7 +125,130 @@ func namespaceLsAction(cmd *cobra.Command, args []string) error {
...
+	client, ctx, cancel, err := newClient(cmd)
+	if err != nil {
+		return err
+	}
+	defer cancel()
+	if err != nil {
+		return err
+	}
+	namespace, labels := ObjectWithLabelArgs(args)
```
Comment (AkihiroSuda) [inline cmd/nerdctl/namespace.go:210]: labels should be taken from `--label`

File: cmd/nerdctl/namespace_linux.go [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+package main
+
+import (
+	"github.com/containerd/containerd/namespaces"
+	"github.com/containerd/containerd/runtime/opts"
+	"github.com/spf13/cobra"
+)
+
+func deleteOpts(cmd *cobra.Command) ([]namespaces.DeleteOpts,error) {
```
Comment (AkihiroSuda) [inline cmd/nerdctl/namespace_linux.go:25]:
```suggestion
func namespaceDeleteOpts(cmd *cobra.Command) ([]namespaces.DeleteOpts,error) {
```

File: README.md [hunk 9]
```diff
@@ -1089,6 +1097,19 @@ Usage: `nerdctl namespace ls [OPTIONS]`
...
+
+Usage: `remove [OPTIONS] NAMESPACE [NAMESPACE...]`
+
+Flags:
+- `-c, --cgroup`: delete the namespace's cgroup
+
+### :nerd_face: :blue_square: nerdctl namespace label
+Set and clear labels for a namespace.
```
Comment (AkihiroSuda) [inline README.md:1109]: How to clear?

File: cmd/nerdctl/namespace.go (hunk 1 above)
Comment (AkihiroSuda) [reply cmd/nerdctl/namespace.go:171]: It should be also documented that we may have dangling objects (e.g., containers, images, snapshots) in the daemon's database when an object was created during we remove the namespace (IIRC)

Comment (ningmingxiao) [reply cmd/nerdctl/namespace.go:168]: i try to delte a not null namespace,it will show error
```
ATA[0000] unable to delete default: namespace "default" must be empty, but it still has images, blobs, containers, snapshot-overlayfs: failed precondition
```

File: cmd/nerdctl/namespace.go (hunk 4 above)
Comment (ningmingxiao) [reply cmd/nerdctl/namespace.go:97]: I refer to "ctr ns ls",do you think is it necessary to add "nerdctl ns inspect" command?
namespace doesn't have many useful message

File: cmd/nerdctl/namespace.go (hunk 7 above)
Comment (ningmingxiao) [reply cmd/nerdctl/namespace.go:205]: now usage is  "nerdctl ns label namespace1 key1=value1",like docker tag command,
i think it is not necessary to use cmd.Flags().GetString("label") to get lable value

File: cmd/nerdctl/namespace.go (hunk 4 above)
Comment (AkihiroSuda) [reply cmd/nerdctl/namespace.go:97]: Yes, for consistency with other commands

File: cmd/nerdctl/namespace.go (hunk 1 above)
Comment (ningmingxiao) [reply cmd/nerdctl/namespace.go:168]: I need  time to find out it

File: README.md (hunk 2 above)
Comment (ningmingxiao) [reply README.md:1090]: like  “nerdctl ns label namespace1  label  key1=value1,there is double lable ,” can you for an example like "nerdctl ns label namespace1 key1=value1" ?

File: cmd/nerdctl/namespace.go (hunk 4 above)
Comment (ningmingxiao) [reply cmd/nerdctl/namespace.go:97]: I will add it tomorrow

File: README.md (hunk 2 above)
Comment (ningmingxiao) [reply README.md:1090]: can you for an example use --label ? @AkihiroSuda

Comment (AkihiroSuda) [reply README.md:1090]:
```
nerdctl namespace create foo --label key1=val1 --label key2=val2

nerdctl namespace update foo --label key3=val3 --label key4=val4
```

File: cmd/nerdctl/namespace.go (hunk 7 above)
Comment (AkihiroSuda) [reply cmd/nerdctl/namespace.go:205]:
```
nerdctl namespace create foo --label key1=val1 --label key2=val2

nerdctl namespace update foo --label key3=val3 --label key4=val4
```

File: cmd/nerdctl/namespace.go (hunk 1 above)
Comment (AkihiroSuda) [reply cmd/nerdctl/namespace.go:168]: `--force` can be added in a separate PR.

File: README.md (hunk 2 above)
Comment (ningmingxiao) [reply README.md]: i use nerdctl namespace update foo  --label "key1 =v1 key2 =v2", cobra doesn't support repeat flag

Comment (AkihiroSuda) [reply README.md]: > cobra doesn't support repeat flag

Untrue. We have `nerdctl run -v vol1 -v vol2` .

Comment (ningmingxiao) [reply README.md]: done,thank you

File: cmd/nerdctl/namespace.go (hunk 1 above)
Comment (ningmingxiao) [reply cmd/nerdctl/namespace.go]: ok

File: cmd/nerdctl/namespace_inspect.go [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+}
+
+func labelInspectAction(cmd *cobra.Command, args []string) error {
+	client, ctx, cancel, err := newClient(cmd)
+	if err != nil {
+		return err
+	}
+	defer cancel()
+
+	result := make([]interface{}, len(args))
```
Comment (AkihiroSuda) [inline cmd/nerdctl/namespace_inspect.go:81]: Why interface{}?

File: cmd/nerdctl/namespace_inspect.go [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+type volumeInfo struct {
+	Name string
+}
+
+type nsInfo struct {
+	Name       string
+	Label      string
+	Containers []containerInfo
+	Images     []imageInfo
+	Volumes    []volumeInfo
```
Comment (AkihiroSuda) [inline cmd/nerdctl/namespace_inspect.go:71]: I’m not sure we want to add Containers, Images, Volumes.
Can be discussed in another PR.

File: cmd/nerdctl/namespace_inspect.go (hunk 10 above)
Comment (ningmingxiao) [reply cmd/nerdctl/namespace_inspect.go:81]: refer to  volumeInspectAction, you mean  result := make([] nsInfo, len(args))?

File: cmd/nerdctl/namespace_inspect.go (hunk 11 above)
Comment (ningmingxiao) [reply cmd/nerdctl/namespace_inspect.go:71]: what  inspect you want to contain?

Comment (AkihiroSuda) [reply cmd/nerdctl/namespace_inspect.go:71]: Just name and labels.

For other objects we may want to reuse https://github.com/containerd/nerdctl/tree/v0.17.1/pkg/inspecttypes/native (in another PR)

File: cmd/nerdctl/namespace_inspect.go [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+	Tag        string
+}
+
+type volumeInfo struct {
+	Name string
+}
+
+type nsInfo struct {
+	Name       string
+	Label      string
```
Comment (AkihiroSuda) [inline cmd/nerdctl/namespace_inspect.go:68]: `Labels     *map[string]string `json:"Labels,omitempty"``

File: cmd/nerdctl/namespace_inspect.go [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+	Id         string
+	Repository string
+	Tag        string
+}
+
+type volumeInfo struct {
+	Name string
+}
+
+type nsInfo struct {
```
Comment (AkihiroSuda) [inline cmd/nerdctl/namespace_inspect.go:66]: Should be defined in https://github.com/containerd/nerdctl/tree/master/pkg/inspecttypes/native

**Review by AkihiroSuda — APPROVED** [review]
Thanks
